app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/ToolbarUiController.kt

```diff
@@ -0,0 +1,158 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.navigationoverlay
+
+import android.arch.lifecycle.LifecycleOwner
+import android.arch.lifecycle.Observer
+import android.support.v4.app.FragmentManager
+import android.view.View
+import android.widget.ImageButton
+import kotlinx.android.synthetic.main.browser_overlay.view.*
+import kotlinx.android.synthetic.main.browser_overlay_top_nav.view.*
+import mozilla.components.browser.domains.autocomplete.ShippedDomainsProvider
+import mozilla.components.support.ktx.android.view.forEach
+import mozilla.components.support.ktx.android.view.hideKeyboard
+import org.mozilla.tv.firefox.ext.forceExhaustive
+import org.mozilla.tv.firefox.ext.serviceLocator
+import org.mozilla.tv.firefox.utils.URLs
+import org.mozilla.tv.firefox.utils.ViewUtils
+import org.mozilla.tv.firefox.widget.IgnoreFocusMovementMethod
+import org.mozilla.tv.firefox.widget.InlineAutocompleteEditText
+
+private const val NAVIGATION_BUTTON_ENABLED_ALPHA = 1.0f
+private const val NAVIGATION_BUTTON_DISABLED_ALPHA = 0.3f
+
+/**
+ * An encapsulation of the toolbar to set up and respond to UI operations.
+ */
+class ToolbarUiController(
```

Comment:
This is an interesting concept and I agree with benefits you've listed; I do question if adding another layer of encapsulation adds more complications to our current architecture. What do you think about creating a custom `ToolbarView` instead? I think it's also worth looking into dependency injection in the near future. @Baron-Severin might also have some good inputs on this as well

Reply:
> I do question if adding another layer of encapsulation adds more complications to our current architecture.

I don't consider this to be an extension of our architecture – it's a grouping of some of the parts of a single architectural component – the NavigationOverlayFragment – necessitated by  the file growing too large. The ToolbarUiController is intended to have same responsibility in the architecture as the Fragment – it's just focused on a specific part of the fragment.

fwiw, the CursorController _kind of_ acts in a similar way: it's a grouping of Cursor functionality that is an extension of WebRenderFragment.

I can see how one might think this is another part of the architecture to understand – especially because it's called "Controller" – and the problems that might entail. I think this could be fixed with a better name but I couldn't come up with one though (e.g. `ToolbarSubFragment` isn't quite right because it isn't really a fragment). I used `Controller` because it evokes a similar responsibility to the Controller in MVC: it (more or less) hooks up the view to the model. Let me know if you can think of a better name (`ToolbarFragmentExtension`?).

> What do you think about creating a custom ToolbarView instead?

I think this would have a larger impact on our architecture: the view would now take on some responsibilities that the fragment current encapsulates and generally our views are pretty dumb.

> I think it's also worth looking into dependency injection in the near future.

This seems unrelated so we can talk about that elsewhere.

Reply:
We can follow-up in a new bug if this discussion makes us decide to change the code.